net.go

```diff
@@ -118,6 +120,233 @@ func (e *EurekaConnection) GetApps() (map[string]*Application, error) {
+	}
+	return count
+}
+
+type instanceQueryOptions struct {
+	// secure indicates whether to query a secure or an insecure VIP address.
+	secure bool
+	// predicate guides filtering, indicating whether to retain an instance when it returns true or
+	// drop it when it returns false.
+	predicate func(*Instance) bool
+	// intn behaves like the rand.Rand.Intn function, aiding in randomizing the order of the result
+	// sequence when non-nil.
+	intn func(int) int
+}
+
+// InstanceQueryOption is a customization supplied to instance query functions like
+// GetInstancesByVIPAddress to tailor the set of instances returned.
+type InstanceQueryOption func(*instanceQueryOptions) error
+
+func retainIfStatusIs(status StatusType, o *instanceQueryOptions) {
+	if prev := o.predicate; prev != nil {
+		o.predicate = func(instance *Instance) bool {
+			return prev(instance) || instance.Status == status
+		}
+	} else {
+		o.predicate = func(instance *Instance) bool {
+			return instance.Status == status
+		}
+	}
+}
+
+// WithStatus restricts the set of instances returned to only those with the given status.
+//
+// Supplying multiple options produced by this function applies their logical disjunction.
+func WithStatus(status StatusType) InstanceQueryOption {
+	return func(o *instanceQueryOptions) error {
+		if len(status) == 0 {
+			return errors.New("invalid instance status")
+		}
+		retainIfStatusIs(status, o)
+		return nil
+	}
+}
+
+func makeUncheckedStatusMatchingOption(status StatusType) InstanceQueryOption {
+	return func(o *instanceQueryOptions) error {
+		retainIfStatusIs(status, o)
+		return nil
+	}
+}
+
+// ThatAreUp restricts the set of instances returned to only those with status UP.
+//
+// Combining the options produced by this function with those produced by WithStatus applies their
+// logical disjunction.
+func ThatAreUp() InstanceQueryOption {
+	return makeUncheckedStatusMatchingOption(UP)
+}
+
+// Shuffled requests randomizing the order of the sequence of instances returned, using the default
+// shared rand.Source.
+func Shuffled() InstanceQueryOption {
+	return func(o *instanceQueryOptions) error {
+		o.intn = rand.Intn
+		return nil
+	}
+}
+
+// ShuffledWith requests randomizing the order of the sequence of instances returned, using the
+// supplied source of random numbers.
+func ShuffledWith(r *rand.Rand) InstanceQueryOption {
+	return func(o *instanceQueryOptions) error {
+		o.intn = r.Intn
+		return nil
+	}
+}
+
+func shuffleInstances(instances []*Instance, intn func(int) int) {
+	count := len(instances)
+	if count < 2 {
+		return
+	}
+	if intn(2) == 0 {
+		instances[1], instances[0] = instances[0], instances[1]
+	}
+	for i := 2; i != count; i++ {
+		if j := intn(i + 1); j != i {
+			instances[i], instances[j] = instances[j], instances[i]
+		}
+	}
+}
+
+func filterInstances(instances []*Instance, pred func(*Instance) bool) []*Instance {
+	for firstBegin, instance := range instances {
+		if !pred(instance) {
+			continue
+		}
```

Comment:
Two days ago I wrote a benchmark for `fargo.filterInstances`, then yesterday took some time to investigate whether eliminating some slice bounds checks (of which there are six in the current implementation) would help speed it up. Revising `fargo.filterInstances` to use range-based loops for the inner two rather than the current index striding style eliminated one bounds check, but didn't reduce the running time.

You can find the benchmark in commit 2e03c507b994780926bd1d2494ccb7c8ef226594. As shown in the commit comment, the current implementation is faster in all cases, and it allocates only when there are at least two sequences of instances to retain separated by a sequence of instances to drop (demonstrated by the "up-down-up" cases).

As you noted, usually the sets of instances are small, so the claimed improvement here doesn't amount to much in those common cases. Please let me know if you'd prefer me to replace this function with the simpler allocate-and-append version. I don't want to preclude _fargo_ learning about VIP addresses because of this function.